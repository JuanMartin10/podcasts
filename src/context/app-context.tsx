import { createContext, useContext, useState } from 'react';
import { useEntries } from '../hooks/useEntries';
import { type AppEntry } from '../models/types';

export interface AppContextTypes {
  loading: boolean;
  entries: AppEntry[] | undefined;
  currentPodcast: AppEntry | undefined;
  getEntries: () => Promise<void>;
  setLoading: React.Dispatch<React.SetStateAction<boolean>>;
  setEntries: React.Dispatch<React.SetStateAction<AppEntry[] | undefined>>;
  setCurrentPodcast: React.Dispatch<React.SetStateAction<AppEntry | undefined>>;
}

interface ContextProps {
  children: React.ReactNode;
}

const AppContext = createContext<AppContextTypes | undefined>(undefined);

export const AppProvider: React.FC<ContextProps> = ({ children }) => {
  const [loading, setLoading] = useState(false);
  const [currentPodcast, setCurrentPodcast] = useState<AppEntry>();

  const { entries, getEntries, setEntries } = useEntries({
    loading,
    setLoading,
  });

  return (
    <AppContext.Provider
      value={{
        loading,
        entries,
        currentPodcast,
        getEntries,
        setLoading,
        setEntries,
        setCurrentPodcast,
      }}
    >
      {children}
    </AppContext.Provider>
  );
};

export const useAppContext = (): AppContextTypes => {
  const context = useContext(AppContext);
  if (context == null) {
    throw new Error('useAppContext must be used within an AppWrapper');
  }
  return context;
};
