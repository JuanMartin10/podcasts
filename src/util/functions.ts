import { Feed } from '../models/api-types';

export const formatTime = (ms: number) => {
  let totalSeconds = Math.floor(ms / 1000);
  const hours = Math.floor(totalSeconds / 3600);
  totalSeconds %= 3600;
  const minutes = Math.floor(totalSeconds / 60);
  const seconds = totalSeconds % 60;
  const secondsString = seconds.toString().padStart(2, '0');

  if (hours > 0) {
    return `${hours}:${minutes.toString().padStart(2, '0')}:${secondsString}`;
  } else {
    return `${minutes}:${secondsString}`;
  }
};

export const formatDate = (dateString: string) => {
  const date = new Date(dateString);

  const year = date.getFullYear();
  const month = (date.getMonth() + 1).toString().padStart(2, '0');
  const day = date.getDate().toString().padStart(2, '0');

  return `${day}/${month}/${year}`;
};

export const formatApiEntries = (apiFeed: Feed) =>
  apiFeed.entry.map(entry => ({
    id: entry.id.attributes['im:id'],
    title: entry['im:name'].label,
    artist: entry['im:artist'].label,
    image: entry['im:image'][0].label,
    summary: entry.summary.label,
  }));
