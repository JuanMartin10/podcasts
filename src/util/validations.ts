export const isEmpty = (val: string | null | undefined) => {
  return val === '' || val === null || val === undefined;
};

export const isFalse = (val: boolean) => !val;
