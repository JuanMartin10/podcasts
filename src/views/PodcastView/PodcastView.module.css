.center {
  display: flex;
  justify-content: center;
  align-items: center;
}
.episodesContainer {
  box-shadow: 0.12em 0.25em 0.5em 0.25em rgba(0, 0, 0, 0.1);
  padding: 16px;
}
.episodesTitle {
  font-size: 1.5rem;
  margin: 0;
}

.episodesList {
  box-shadow: 0.12em 0.25em 0.5em 0.25em rgba(0, 0, 0, 0.1);
  list-style-type: none;
  padding: 0;
  margin: 24px 0;
}

th {
  text-align: left;
  padding: 8px;
}

.table td {
  padding: 8px;
}
.table tr:nth-child(odd) {
  background: #f1f1f1;
}

.table tr:nth-child(even) {
  background: #fff;
}

.episodeItem {
  margin-bottom: 0.5rem;
}

.episodeTitle {
  font-size: 1.2rem;
  margin-bottom: 0.2rem;
}

.episodeDate {
  font-size: 1rem;
  color: #666;
}
