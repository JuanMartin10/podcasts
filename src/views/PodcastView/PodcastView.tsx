import { useEffect, useState } from 'react';
import styles from './PodcastView.module.css';
import { Link, useParams } from 'react-router-dom';
import { type ResultPodcast } from '../../models/api-podcast.types';
import DetailsLayout from '../../layout/DetailsLayout/DetailsLayout';
import { useAppContext } from '../../context/app-context';
import { fetchEpisodesByPodcast } from '../../services';
import { isEmpty } from '../../util/validations';
import { formatDate, formatTime } from '../../util/functions';

const PodcastView = () => {
  const { entries, currentPodcast, setLoading, setCurrentPodcast } =
    useAppContext();
  const [episodes, setEpisodes] = useState<ResultPodcast[] | undefined>();
  const { podcastId } = useParams();

  useEffect(() => {
    setLoading(true);
    const podcastFromStorage = localStorage.getItem(`podcast_${podcastId}`);
    if (entries == null) return;
    if (podcastFromStorage == null) {
      const p = entries?.find((el: any) => el.id === podcastId?.toString());
      const objToSave = { podcast: p };
      fetchEpisodesByPodcast(podcastId as string).then(res => {
        setCurrentPodcast(p);
        setEpisodes(res);
        const obj = { ...objToSave, episodes: res };
        localStorage.setItem(`podcast_${podcastId}`, JSON.stringify(obj));
        setLoading(false);
      });
    } else {
      const parse = JSON.parse(podcastFromStorage);
      setCurrentPodcast(parse.podcast);
      setEpisodes(parse.episodes);
      setLoading(false);
    }
  }, [entries]);

  useEffect(() => {}, [currentPodcast]);

  return isEmpty(currentPodcast as undefined) ? (
    <div className={styles.center}>
      <p>Cargando datos...</p>
    </div>
  ) : (
    <DetailsLayout currentPodcast={currentPodcast}>
      <div className={styles.episodesContainer}>
        <h3 className={styles.episodesTitle}>Episodes: {episodes?.length}</h3>
      </div>
      <div className={styles.episodesList}>
        <table width='100%'>
          <thead>
            <tr>
              <th>Title</th>
              <th>Date</th>
              <th>Duration</th>
            </tr>
          </thead>
          <tbody className={styles.table}>
            {episodes != null &&
              episodes?.length > 0 &&
              episodes.map((episode: any) => (
                <tr key={`${episode.trackId} ${episode.trackName}`}>
                  <td>
                    <Link
                      to={`/podcast/${podcastId}/episode/${episode.trackId}`}
                      state={episode}
                    >
                      {episode.trackName}
                    </Link>
                  </td>
                  <td>{formatDate(episode.releaseDate)}</td>
                  <td>{formatTime(episode.trackTimeMillis)}</td>
                </tr>
              ))}
          </tbody>
        </table>
      </div>
    </DetailsLayout>
  );
};

export default PodcastView;
