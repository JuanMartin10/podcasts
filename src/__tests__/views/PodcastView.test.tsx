import { afterEach, beforeEach, describe, expect, it } from 'vitest';
import { render, screen, cleanup, waitFor } from '@testing-library/react';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { AppContext } from 'context/app-context';
import currentPodcastJson from 'mocks/podcast.json';
import episodesJson from 'mocks/episodes.json';
import { formatApiEntries } from 'util/functions';
import PodcastView from 'views/PodcastView/PodcastView';

const renderWithProvider = (ui: React.ReactElement) =>
  render(
    <AppContext.Provider
      value={{
        entries: formatApiEntries(episodesJson.feed as any),
        loading: false,
        currentPodcast: currentPodcastJson.podcast,
        setLoading: () => {},
        setCurrentPodcast: () => {},
        setEntries: () => {},
      }}
    >
      {ui}
    </AppContext.Provider>
  );
const renderLoading = (ui: React.ReactElement) =>
  render(
    <AppContext.Provider
      value={{
        entries: formatApiEntries(episodesJson.feed as any),
        loading: true,
        currentPodcast: undefined,
        setLoading: () => {},
        setCurrentPodcast: () => {},
        setEntries: () => {},
      }}
    >
      {ui}
    </AppContext.Provider>
  );

const wrapper = ({ id }: any) => (
  <MemoryRouter initialEntries={[`/podcast/${id}`]}>
    <Routes>
      <Route path='/podcast/:id' Component={PodcastView} />
    </Routes>
  </MemoryRouter>
);

describe('<PodcastView />', () => {
  const wrapperId = () => wrapper({ id: '1574029840' });
  beforeEach(() => {
    renderWithProvider(wrapperId());
  });
  afterEach(cleanup);

  it('Should render', () => {
    renderWithProvider(wrapperId());
  });

  it('Should match the snapshot', async () => {
    const { container } = renderWithProvider(wrapperId());
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });

  it('Should render Friday Night Karaoke as h2', () => {
    const text = screen.getAllByText(/friday night karaoke/i);
    const h2Element = text.find(element => element.tagName === 'H2');

    expect(h2Element).toBeDefined();
  });

  it("Should render 'Loading...'", async () => {
    renderLoading(wrapperId());
    screen.getByText(/loading/i);
    await waitFor(() => {
      expect(screen.getByText(/loading/i)).toBeDefined();
    });
  });

  it('Should render podcast image', async () => {
    await waitFor(() => {
      const images = screen.getAllByRole('img');
      expect(images).toHaveLength(1);
      expect(images).toBeDefined();
    });
  });

  it('Should render description', () => {
    waitFor(() => {
      const description = screen.getByText(
        /No ads, no gimmicks - just Karaoke! Friday Night/i
      );
      expect(description).toBeDefined();
    });
  });

  it('Should render 53 episodes', async () => {
    waitFor(() => {
      const allEpisodes = screen.getAllByRole('tr');
      expect(allEpisodes).toHaveLength(53);
      const episode = screen.getByText('Hit Me With Elvis');
      expect(episode).toBeDefined();
    });
  });
});
