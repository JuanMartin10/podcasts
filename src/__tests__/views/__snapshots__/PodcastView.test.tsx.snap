// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<PodcastView /> > Should match the snapshot 1`] = `
<div>
  <div
    class="_container_453f42"
  >
    <div
      class="_sidebar_b37e70"
    >
      <a
        class="_link_b37e70"
        href="/podcast/1574029840"
      >
        <img
          alt="Friday Night Karaoke"
          class="_image_b37e70"
          src="https://is1-ssl.mzstatic.com/image/thumb/Podcasts116/v4/5b/88/a1/5b88a186-ce6a-5268-cc49-8e896c737729/mza_7354436562524572096.jpg/55x55bb.png"
        />
      </a>
      <div
        class="_descriptionContainer_b37e70"
      >
        <a
          class="_link_b37e70"
          href="/podcast/1574029840"
        >
          <h2
            class="_title_b37e70"
          >
            Friday Night Karaoke
          </h2>
          <p
            class="_author_b37e70"
          >
            <span>
              by
            </span>
             
            Friday Night Karaoke
          </p>
        </a>
      </div>
      <div
        class="_descriptionContainer_b37e70"
      >
        <p>
          Description:
        </p>
        <p
          class="_description_b37e70"
        >
          No ads, no gimmicks - just Karaoke! Friday Night Karaoke features amateur artists every week singing the songs they love, just for you! Get your weekly dose of vocal expression. Get featured on the podcast by joining the official Friday Night Karaoke Facebook group at https://www.facebook.com/groups/fridaynightkaraoke!
        </p>
      </div>
    </div>
    <div
      class="_main_453f42"
    >
      <div
        class="_episodesContainer_4cfd7d"
      >
        <h3
          class="_episodesTitle_4cfd7d"
        >
          Episodes: 
        </h3>
      </div>
      <div
        class="_episodesList_4cfd7d"
      >
        <table
          width="100%"
        >
          <thead>
            <tr>
              <th>
                Title
              </th>
              <th>
                Date
              </th>
              <th>
                Duration
              </th>
            </tr>
          </thead>
          <tbody
            class="_table_4cfd7d"
          />
        </table>
      </div>
    </div>
  </div>
</div>
`;
